r/R/io.R

```diff
@@ -0,0 +1,73 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#' @include R6.R
+#' @include enums.R
+#' @include buffer.R
+
+`arrow::io::InputStream` <- R6Class("arrow::io::InputStream", inherit = `arrow::Object`)
+
+`arrow::io::RandomAccessFile` <- R6Class("arrow::io::RandomAccessFile", inherit = `arrow::io::InputStream`,
+  public = list(
+    GetSize = function() io___RandomAccessFile__GetSize(self),
+    supports_zero_copy = function() io___RandomAccessFile__supports_zero_copy(self)
+  )
+)
+
+`arrow::io::MemoryMappedFile` <- R6Class("arrow::io::MemoryMappedFile", inherit = `arrow::io::RandomAccessFile`,
+  public = list(
+    Close = function() io___MemoryMappedFile__Close(self),
+    Tell = function() io___MemoryMappedFile__Tell(self),
+    Seek = function(position) io___MemoryMappedFile__Seek(self, position),
+    Resize = function(size) io___MemoryMappedFile__Resize(self, size),
+    Read = function(nbytes) `arrow::Buffer`$new(io___Readable__Read(self, nbytes))
```

Comment:
Does R6 support diamond inheritance? You probably want to follow the Arrow class hierarchy and put `Close`, `Tell`, and `Seek` in a base class, and `Read` on `InputStream`

Reply:
I don't think so, ping @wch.  https://github.com/r-lib/R6/issues/9

Reply:
I see. Well you could implement a non-diamond inheritance hierarchy for R and at least maximize the amount of code sharing. I guess there might be the issue that the incoming shared_ptr type to each function is different, so perhaps it's not worth the effort as long as you internally dispatch to a primary implementation that uses the lowest-level virtual interface

Reply:
That's right, R6 doesn't have multiple inheritance.

Reply:
Would something like java's `implements` make sense ?

Reply:
@romainfrancois want to file an issue on R6? It's worth having a discussion about it.